src/python/gem5/components/processors/simpoint.py

```diff
@@ -0,0 +1,171 @@
+# Copyright (c) 2021 The Regents of the University of California
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met: redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer;
+# redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution;
+# neither the name of the copyright holders nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from m5.util import fatal
+from pathlib import Path
+from typing import List, Tuple
+
+
+class SimPoint:
+    """
+    This SimPoint class is used to manage the information needed for SimPoints
+    in workload
+
+    """
+
+    def __init__(
+        self,
+        simpoint_interval: int,
+        simpoint_file_path: Path = None,
+        weight_file_path: Path = None,
+        simpoint_list: List[int] = None,
+        weight_list: List[int] = None,
+        warmup_interval: int = 0
+        # bbv_file_path: Path = None, will take in bbv and generate simpoints
+        # with it in the future
+    ) -> None:
+        """
+        :param simpoint_interval: the length of each SimPoints interval
+        :param simpoint_file_path: the path to the SimPoints result file
+        generated by Simpoint3.2 or gem5
+        :param weight_file_path: the path to the weight result file generated
+        by Simpoint3.2 or gem5
+
+        :param simpoint_list: a list of SimPoints starting instructions
+        :param weight_list: a list of SimPoints weights
+        :param warmup_interval: a number of instructions for warming up before
+        restoring a SimPoints checkpoint
+
+        usage note
+        -----------
+        Need to pass in the paths or the lists for the SimPoints and their
+        weights. If the paths are passed in, no actions will be done to the
+        list.
+
```

Comment:
Perhaps given this usage there should be two `def __init__` functions: one which accepts paths, and another that accepts lists. Is this possible to do? It'd make it much clearer to the user how this is used.

Reply:
I'll push back here. Python doesn't support differentiating functions on input types (it's duck typed!). So, I think this is actually a great way to do it.

You *could* have two function `_init_lists()` and `_init_files()` which the `__init__` calls, but I'm not sure that's any better.

Reply:
Fair, was just wondering if there was a nice way to do it. If not then go with this.